### PR TITLE
Bls operations rebased

### PR DESCRIPTION
**Pull request by nano-o**

### CODE REVIEW

File: proof/bls_operations.saw [hunk 1]
```diff
@@ -0,0 +1,1093 @@
...
+                    , msg_ptr
+                    , llvm_term {{ `msg_len : [64] }} ];
+  // TODO: Endianness?  (Both for return, and for args to CoreSign)
+  llvm_points_to
+    out_ptr
+    (llvm_term {{ split`{each=8} (CoreSign (drop_join_reverse`{255, 32} SK,
+                                            msg)) }});
+};
+
+// TODO: Under what conditions is this true?  Should probably be done over IETF
```
Comment (nano-o) [inline proof/bls_operations.saw:452]: I think some version of this is proved using an assumption about the high-level spec in `hash_to_g2.saw`

File: proof/bls_operations.saw [hunk 2]
```diff
@@ -0,0 +1,1093 @@
...
+hash_to_curve_arg_expand_thm <- prove_cryptol
+  {{ \(x:([32][8],[43][8])) ->
+       hash_to_curve_opt_impl x ==
+       apply hash_to_curve_opt_impl (([]:[0][8])#x.0, x.1) }} ["hash_to_curve_opt_impl"];
+
+// TODO: why is this needed here?
+hash_to_curve_E1_opt_impl_equiv_ov <- (prove_hash_to_curve_E1_opt_impl_equiv_thm 32 43);
+
+// TODO: why is this needed here (see prove_hash_to_g1_ov in hash_to_g1.saw)?
+hash_to_curve_e1_impl_thm <- custom_prove_cryptol
```
Comment (nano-o) [inline proof/bls_operations.saw:564]: Could we use `blst_hash_to_g1_spec` instead so that we get the high-level spec directly?

Comment (bboston7) [reply proof/bls_operations.saw:557]: Maybe?  The reason I used `blst_hash_to_g1_impl_spec` is that it's postcondition is written with `llvm_points_to`, while `blst_hash_to_g1_spec` is written with `llvm_postcond`.  The `llvm_postcond` forms tend to be painful to work with.

**Comment by nano-o** [comment]
@bboston7 can you have a look before we merge it into main?

**Review by bboston7 — APPROVED** [review]
I think this is OK to merge.  I will clean up the BLS proof wrappers in a future PR.

File: proof/core_verify_pk_in_g1.saw [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+// NOTE: calling the function with a null pk is not safe
+
+let {{
+  blst_core_verify_pk_in_g1_error_precond : POINTonE1_affine_t -> POINTonE2_affine_t -> Bool
+  blst_core_verify_pk_in_g1_error_precond pk sig =
+       is_point_O E (POINTonE1_affine_abs pk) \/ ( ~(is_point_O E' (POINTonE2_affine_abs sig)) /\ ~is_in_g2_impl (POINTonE2_affine_abs sig) )
+}};
+
+// case in which no argument is null and the error condition is false
+let blst_core_verify_pk_in_g1_non_null_spec msg_len dst_len aug_len = do {
```
[the remark this answers is not in the document]
Comment (nano-o) [reply proof/core_verify_pk_in_g1.saw:29]: okay, will do that in a later PR

**Pull request merged**